enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,467 @@
+Prometheus monitoring stack. 
+
+On the other hand, data aggregation will be required using the metrics API, as
+decisions about scaling nodes up and down, must not be performed based on
+performance usage at a single point in time. Instead the operator will have to
+gather and store usage data for a certain period of time, to decide if scaling
+is required.
+
+As an example with a configured `timeWindow` (see
+[configuration](#control-plane-machineset-operator-configuration)) of 30 minutes
+for scaling up, the operator has to keep querying the metrics API and averaging
+values until enough data has been aggregated to know the load average for the
+last 30 minutes.
+
+If the metrics API returns it's own window of ~60 second aggregations, the
+operators will have to retrieve 30 values (one per minute), before any scaling
+decision should be made.
+
+#### Using Prometheus
+
+Using Prometheus generated data will make the implementation easier with respect
+to data aggregation, as data should already have been aggregated by prometheus.
+
+#### Hypershift
+
+Hypershift does not use `ControlPlaneMachineSet` to manage the control-planes
+for hosted clusters, as those are not running on their own virtual machines.
+
+Because of this, no special handling should be required.
+
+### Risks and Mitigations
+
+#### Risk 1: Higher costs when scaling up
+
+Scaling up will increase costs for running the cluster, as bigger nodes will
+incur higher costs from the cloud provider.
+
+To mitigate unbounded scaling the user can setup the possible instance sizes
+available for scaling - once the highest size has been scaled to, no more
+scaling will occur.
+
+#### Risk 2: Cluster instability due to scaling down
+
+Scaling down might also impact the clusters ability to accommodate a workload if
+resource usage was uncharacteristically low for the detection period of scaling.
+This turns into an issue, if the cluster still would have required the bigger
+control plane nodes.
+
+To mitigate this issue the scale down behavior should be disabled by default and
+require an explicit opt-in by the cluster administrator.
+
+#### Risk 3: Control plane node churn
+
+Automatically scaling up and down control planes could impact cluster
+reliability and availability, if performed too often.
+
+As such the additional logic must ensure that there is a grace period between
+changes to the machine size, during which no more scaling will be performed.
+
+### Drawbacks
+
+Some users might not want this feature and rather opt to scale control planes
+manually.
+
+## Design Details
+
+### Control Plane MachinSet operator configuration
+
+As automatic scaling will require adjustments to certain aspects of its logic,
+the implementation should implement a way for administrators of an OpenShift
+cluster to configure the specifics of how and when scaling will occur.
+
+This configuration will be performed using a new custom resource
```

Comment:
I like the idea yes, the operator in general should probably be a singleton, but not sure if/how this can be enforced.
While the kubernetes API will prevent any races in the modification to the CPMS CR itself, having 2 operators with completely different configurations for when to scale could result in a lot of churn.

Reply:
With CPMS, we use a validation to ensure that the name of the resource can only ever be `cluster`. I would probably expect the same for this and assert that there is a 1:1 mapping between the configuration and the CPMS it looks after, so, you'd only expect one to be created in the `openshift-machine-api`, or in the future, `openshift-cluster-api` namespaces

Reply:
Hm while checking this comment, it seems to me that the PR didn't reflect the latest push I made 2 weeks ago - I already added that I expect this CR to be a singleton to prevent the possibility of having the operator run two times and then racing it's own resizes.